### PR TITLE
Support specialization constants.

### LINKED ISSUE #3581: Shader constants not working?
I want to use a shader with a shader constant like this

```glsl
layout(constant_id = 0) const int MULTIPLIER = 0;
```

but the code generated by slang always wants to use a uniform descriptor at Set 0 Binding 1. Is this not implemented or is this a bug? I am using vulkan.

edit: Not talking about the slang feature #3565 but the functionality used in vulkan, slang just needs to generate spirv that allows this

### PR DESCRIPTION
**Pull request by csyonghe**
This PR adds support for specialization constants in both spirv and glsl backends.

Speecialization constants can now be declared with:
```
[vk::specialization_constant]
const int specConst = 1; // implicit layout

[vk::constant_id(7)] // explicit layout
const int specConst = 1;

layout(constant_id = 9]
const int specConst = 1; // glsl flavor
```

Closes #3581.

### CODE REVIEW

**Review by cheneym2 — APPROVED** [review]
LGTM

**Review by jkwak-work — COMMENTED** [review]
Looks good to me.

Sorry for my late reply.

**Comment by smarton-sc** [comment]
@csyonghe I just glanced at the code and I hope I didn't completely miss the ball, but it doesn't seem like this is the solution folks need. We don't just need specialization constants for the vulkan backend. It needs to work for all backends. Specialization constants need to be first class citizens in slang, and be translated to all backends appropriately (specialization constants for vulkan, function constants for metal, #ifdef for glsl, etc.). The same way that SPIRV-Cross translates OpSpecConstant to all backends. 

Having specilization at slang link time is cute, but it doesn't solve the problem of exploding shader permutations and compile times on target platforms. We need to emit platform shaders that are specializable natively for those reasons. Ex: compiled metallib with runtime specializable function constants.

Slang is not meant to be run at runtime for the most part, it should be an offline compiler, so giving it "efficient" specialization at the slang IR level is almost irrelevant. The platform shader that Slang outputs is the one that needs to be specializable efficiently at runtime. At runtime we only want to be running the lowest level platform compiler that's available (ex: create MTLFunction with newFunctionWithName from MTLLibrary while specifying MTLFunctionConstantValues). No other overhead, and no source code of any kind should be seen at runtime (on platforms that don't require source, ie: all but GL and WebGPU - the lost APIs). 

To end on a high note, I LOVE that you added implicit constant id, as that's indispensable and it was going to be my second request :) 

I would love to use Slang in production, as all other decisions made by the language seem great! But, without efficient runtime specialization, it's not usable.

**Comment by csyonghe** [comment]
Thank you for your feedback! I agree that we should add support for it on other targets, this shouldn’t be hard to do. Can you file another issue for this request?

**Comment by smarton-sc** [comment]
Will do. Thanks!

File: tests/spirv/specialization-constant.slang [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const bool constValue2 = true;
+
+layout(constant_id = 9) const int constValue3 = 111;
+
+RWStructuredBuffer<float> outputBuffer;
+
+[numthreads(1, 1, 1)]
+void computeMain()
+{
+    if (constValue2)
```
Comment (smarton-sc) [inline tests/spirv/specialization-constant.slang:44]: @csyonghe can we please add a more complex expression here, like "if (constValue2 && constValue0 >= 1 && uni0 >= 1)" to make sure spec const expressions are working, even when mixed with non-spec-const variables? Also, maybe something like:

float arr[constValue0+1];
for (int i = 0; i< constValue0; ++i)
{
   arr[i] = (i < constValue0 ? i : i*i);
}
return arr[int(constValue1)/2];

Comment (csyonghe) [reply tests/spirv/specialization-constant.slang:44]: Yes, we can definitely enhance the test with more complex expressions. Is this something you would like to submit a patch for?

Comment (smarton-sc) [reply tests/spirv/specialization-constant.slang:44]: I could, but probably not anytime soon :(

**Pull request merged**